ONSdigital/census-rm-pubsub-adapter
# Quarantine messages we can't unmarshal in rabbit

# Motivation and Context
We need to persist messages that error somewhere so they can be alerted on/dealth with appropriately later.
Also includes a refactor to call the message processor synchronously from the pubsub receiver, as recommended by [their documentation](https://godoc.org/cloud.google.com/go/pubsub#Subscription.Receive) since the receive function handles concurrency itself. As it was, we were funneling that concurrency down into a single processing function creating blockages rather than assuaging them. 

# What has changed
* Add quarantine queue, send messages we can't unmarshal or validate
* Refactor processor function to be called synchronously by pubsub receiver
* Improve contextual logging within the message processors
* Add validation functions to each pubsub message type
* Modify docker file so depenencies can be a cached layer

# How to test?
Run the service against it's docker compose dependencies, try posting and invalid message in. It should now log an error and send it to the quarantine queue.
Run the acceptance tests with pubsub adapter running instead of pubsubsvc.

# Links
https://trello.com/c/y4rKo0fl/906-pubsub-adapter-dlq-messages-it-cant-process-8
https://godoc.org/cloud.google.com/go/pubsub#Subscription.Receive
